File: prob7.go
package main

import "fmt"

func main() {
	nth := 10001
	n := 0

	for i := 1;; i++ {
		if isPrime(i) {
			n++

			if n == nth {
				fmt.Println(i)
				return
			}
		}
	}
}

func isPrime(a int) bool {
	// check whether a is prime
	if (a <= 1) { return false }
	if (a == 2) { return true }
	if (a % 2 == 0) { return false }

	for i := 3; i < a; i += 2 {
		if (a % i == 0) { return false }
		if (i*i > a) { break }
	}

	return true
}

File: prob7.clj
;; By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we
;; can see that the 6th prime is 13.

;; What is the 10 001st prime number?

(defn is-prime? [num]
  ;; primality test by trail-division
  (cond (< num 2) false
        (= num 2) true
        (even? num) false
        :else (loop [n 3]
                (cond (> (* n n) num) true
                      (zero? (rem num n)) false
                      :else (recur (+ 2 n))))))

(defn primes []
  ;; generate lazy seq of primes
  (let [next-primes (fn lazy-primes [num]
                      ;; return lazy prime primes starting from num
                      (if (is-prime? num)
                        (cons num (lazy-seq (lazy-primes (inc num))))
                        (lazy-seq (lazy-primes (inc num)))))]
    (next-primes 2)))

(defn prob7 []
  (nth (primes) (dec 10001)))
